File: decoders/DL-ISF/DL-ISF.erl

% https://www.decentlab.com/products/sapflow-sensor-for-lorawan

-module(decentlab_decoder).
-define(PROTOCOL_VERSION, 2).
-export([decode/1, start/0]).

-define(SENSOR_DEFS,
  [
    #{
      length => 16,
      values => [
        #{
          name => <<"Sap flow">>,
          convert => fun(X) -> (lists:nth(0 + 1, X) * 16 - 50000) / 1000 end,
          unit => <<"l⋅h⁻¹"/utf8>>
        },
        #{
          name => <<"Heat velocity (outer)">>,
          convert => fun(X) -> (lists:nth(1 + 1, X) * 16 - 50000) / 1000 end,
          unit => <<"cm⋅h⁻¹"/utf8>>
        },
        #{
          name => <<"Heat velocity (inner)">>,
          convert => fun(X) -> (lists:nth(2 + 1, X) * 16 - 50000) / 1000 end,
          unit => <<"cm⋅h⁻¹"/utf8>>
        },
        #{
          name => <<"Alpha (outer)">>,
          convert => fun(X) -> (lists:nth(3 + 1, X) * 32 - 1000000) / 100000 end
        },
        #{
          name => <<"Alpha (inner)">>,
          convert => fun(X) -> (lists:nth(4 + 1, X) * 32 - 1000000) / 100000 end
        },
        #{
          name => <<"Beta (outer)">>,
          convert => fun(X) -> (lists:nth(5 + 1, X) * 32 - 1000000) / 100000 end
        },
        #{
          name => <<"Beta (inner)">>,
          convert => fun(X) -> (lists:nth(6 + 1, X) * 32 - 1000000) / 100000 end
        },
        #{
          name => <<"Tmax (outer)">>,
          convert => fun(X) -> (lists:nth(7 + 1, X) * 2) / 1000 end,
          unit => <<"s"/utf8>>
        },
        #{
          name => <<"Tmax (inner)">>,
          convert => fun(X) -> (lists:nth(8 + 1, X) * 2) / 1000 end,
          unit => <<"s"/utf8>>
        },
        #{
          name => <<"Temperature (outer)">>,
          convert => fun(X) -> (lists:nth(9 + 1, X) - 32768) / 100 end,
          unit => <<"°C"/utf8>>
        },
        #{
          name => <<"Max voltage">>,
          convert => fun(X) -> (lists:nth(10 + 1, X) - 32768) / 1000 end,
          unit => <<"V"/utf8>>
        },
        #{
          name => <<"Min voltage">>,
          convert => fun(X) -> (lists:nth(11 + 1, X) - 32768) / 1000 end,
          unit => <<"V"/utf8>>
        },
        #{
          name => <<"Diagnostic">>,
          convert => fun(X) -> lists:nth(12 + 1, X) + lists:nth(13 + 1, X) * 65536 end
        },
        #{
          name => <<"Upstream Tmax (outer)">>,
          convert => fun(X) -> (lists:nth(14 + 1, X) * 2) / 1000 end,
          unit => <<"s"/utf8>>
        },
        #{
          name => <<"Upstream Tmax (inner)">>,
          convert => fun(X) -> (lists:nth(15 + 1, X) * 2) / 1000 end,
          unit => <<"s"/utf8>>
        }
      ]
    },
    #{
      length => 1,
      values => [
        #{
          name => <<"Battery voltage">>,
          convert => fun(X) -> lists:nth(0 + 1, X) / 1000 end,
          unit => <<"V"/utf8>>
        }
      ]
    }
  ]
).


decode(<<?PROTOCOL_VERSION, DeviceId:16,
         Flags:16/bitstring, Bytes/binary>> = Binary) when is_binary(Binary) ->
  Words = bytes_to_words(Bytes),
  maps:merge(#{<<"Protocol version">> => ?PROTOCOL_VERSION,
               <<"Device ID">> => DeviceId},
             sensor(Words, Flags, ?SENSOR_DEFS));
decode(HexStr) ->
  Binary = hex2bin(HexStr),
  decode(Binary).


start() ->
  io:format("~p~n", [decode("023d0100030c290bab0c3e79707a1d78437991490845997e4cacdeaa6e00000000457e415a0b59")]),
  io:format("~p~n", [decode("023d0100020b59")]).


%%%%%%%%%%%%%%%%%%%%%%%%%%%%% PRIVATE FUNCTIONS %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

sensor(_Words, _Flags, []) -> #{};
sensor(Words, <<Flags:15, 0:1>>, [_Cur | Rest]) ->
  sensor(Words, <<0:1, Flags:15>>, Rest);
sensor(Words, <<Flags:15, 1:1>>,
       [#{length := Len, values := ValueDefs} | RestDefs]) ->
  {X, RestWords} = lists:split(Len, Words),
  Values = value(ValueDefs, X),
  RestSensors = sensor(RestWords, <<0:1, Flags:15>>, RestDefs),
  maps:merge(Values, RestSensors).


value([], _X) -> #{};
value([#{convert := Convert,
         name := Name} = ValueDef | RestDefs], X) ->
  Values = value(RestDefs, X),
  Value = #{<<"value">> => Convert(X)},
  Value2 = add_unit(Value, ValueDef),
  maps:put(Name, Value2, Values);
value([_NoConvert | RestDefs], X) -> value(RestDefs, X).


add_unit(Value, #{unit := Unit}) ->
  maps:put(<<"unit">>, Unit, Value);
add_unit(Value, _NoUnit) ->
  Value.


hex2bin(HexStr) when is_binary(HexStr) -> hex2bin(binary_to_list(HexStr));
hex2bin(HexStr) -> hex2bin(HexStr, []).

hex2bin("", Bins) -> list_to_binary(lists:reverse(Bins));
hex2bin(HexStr, Bins) ->
  Bytes = string:slice(HexStr, 0, 2),
  Word = list_to_integer(Bytes, 16),
  hex2bin(string:slice(HexStr, 2), [Word | Bins]).


bytes_to_words(Bytes) -> bytes_to_words(Bytes, []).

bytes_to_words(<<>>, Words) -> lists:reverse(Words);
bytes_to_words(<<Word:16, Rest/binary>>, Words) ->
  bytes_to_words(Rest, [Word | Words]).

where(true, IfTrue, _IfFalse) -> IfTrue;
where(false, _IfTrue, IfFalse) -> IfFalse.

File: decoders/DL-ISF/DL-ISF.ex

# https://www.decentlab.com/products/sapflow-sensor-for-lorawan

defmodule DecentlabDecoder do
  @protocol_version 2

  defp sensor_defs do
    [
      %{
        length: 16,
        values: [
          %{
            :name => "Sap flow",
            :convert => fn x -> (Enum.at(x, 0) * 16 - 50000) / 1000 end,
            :unit => "l⋅h⁻¹"
          },
          %{
            :name => "Heat velocity (outer)",
            :convert => fn x -> (Enum.at(x, 1) * 16 - 50000) / 1000 end,
            :unit => "cm⋅h⁻¹"
          },
          %{
            :name => "Heat velocity (inner)",
            :convert => fn x -> (Enum.at(x, 2) * 16 - 50000) / 1000 end,
            :unit => "cm⋅h⁻¹"
          },
          %{
            :name => "Alpha (outer)",
            :convert => fn x -> (Enum.at(x, 3) * 32 - 1000000) / 100000 end,
            :unit => nil
          },
          %{
            :name => "Alpha (inner)",
            :convert => fn x -> (Enum.at(x, 4) * 32 - 1000000) / 100000 end,
            :unit => nil
          },
          %{
            :name => "Beta (outer)",
            :convert => fn x -> (Enum.at(x, 5) * 32 - 1000000) / 100000 end,
            :unit => nil
          },
          %{
            :name => "Beta (inner)",
            :convert => fn x -> (Enum.at(x, 6) * 32 - 1000000) / 100000 end,
            :unit => nil
          },
          %{
            :name => "Tmax (outer)",
            :convert => fn x -> (Enum.at(x, 7) * 2) / 1000 end,
            :unit => "s"
          },
          %{
            :name => "Tmax (inner)",
            :convert => fn x -> (Enum.at(x, 8) * 2) / 1000 end,
            :unit => "s"
          },
          %{
            :name => "Temperature (outer)",
            :convert => fn x -> (Enum.at(x, 9) - 32768) / 100 end,
            :unit => "°C"
          },
          %{
            :name => "Max voltage",
            :convert => fn x -> (Enum.at(x, 10) - 32768) / 1000 end,
            :unit => "V"
          },
          %{
            :name => "Min voltage",
            :convert => fn x -> (Enum.at(x, 11) - 32768) / 1000 end,
            :unit => "V"
          },
          %{
            :name => "Diagnostic",
            :convert => fn x -> Enum.at(x, 12) + Enum.at(x, 13) * 65536 end,
            :unit => nil
          },
          %{
            :name => "Upstream Tmax (outer)",
            :convert => fn x -> (Enum.at(x, 14) * 2) / 1000 end,
            :unit => "s"
          },
          %{
            :name => "Upstream Tmax (inner)",
            :convert => fn x -> (Enum.at(x, 15) * 2) / 1000 end,
            :unit => "s"
          }
        ]
      },
      %{
        length: 1,
        values: [
          %{
            :name => "Battery voltage",
            :convert => fn x -> Enum.at(x, 0) / 1000 end,
            :unit => "V"
          }
        ]
      }
    ]
  end


  def decode(msg, :hex) do
    {:ok, bytes} = Base.decode16(msg, case: :mixed)
    decode(bytes)
  end

  def decode(msg) when is_binary(msg), do: decode_binary(msg)

  def decode(msg), do: to_string(msg) |> decode

  defp decode_binary(<<@protocol_version, device_id::size(16), flags::binary-size(2), bytes::binary>>) do
    bytes
    |> bytes_to_words()
    |> sensor(flags, sensor_defs())
    |> Map.put("Device ID", device_id)
    |> Map.put("Protocol version", @protocol_version)
  end


  defp bytes_to_words(<<>>), do: []

  defp bytes_to_words(<<word::size(16), rest::binary>>), do: [word | bytes_to_words(rest)]


  defp sensor(words, <<flags::size(15), 1::size(1)>>, [%{length: len, values: value_defs} | rest]) do
    {x, rest_words} = Enum.split(words, len)
    value(value_defs, x)
    |> Map.merge(sensor(rest_words, <<0::size(1), flags::size(15)>>, rest))
  end

  defp sensor(words, <<flags::size(15), 0::size(1)>>, [_cur | rest]) do
    sensor(words, <<0::size(1), flags::size(15)>>, rest)
  end

  defp sensor([], _flags, []), do: %{}


  defp value([], _x), do: %{}

  defp value([%{convert: nil} | rest], x), do: value(rest, x)

  defp value([%{name: name, unit: unit, convert: convert} | rest], x) do
    value(rest, x)
    |> Map.put(name, %{"unit" => unit, "value" => convert.(x)})
  end

  defp where(true, if_true, _if_false), do: if_true
  defp where(false, _if_true, if_false), do: if_false

end

IO.inspect(DecentlabDecoder.decode("023d0100030c290bab0c3e79707a1d78437991490845997e4cacdeaa6e00000000457e415a0b59", :hex))
IO.inspect(DecentlabDecoder.decode("023d0100020b59", :hex))
